<a href="https://colab.research.google.com/github/abdullahkrs/e-Portfolio-Machine-Learning-January-2025/blob/main/Jaccard_Coefficient_Pathology.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# 🤝 Jaccard Coefficient Calculations – Pathological Test Results

This notebook calculates the **Jaccard Coefficient** between pairs of individuals based on their symptom and test results.

## 📘 Input Table

| Name | Gender | Fever | Cough | Test-1 | Test-2 | Test-3 | Test-4 |
|------|--------|-------|-------|--------|--------|--------|--------|
| Jack | M      | Y     | N     | P      | N      | N      | A      |
| Mary | F      | Y     | N     | P      | A      | P      | N      |
| Jim  | M      | Y     | P     | N      | N      | N      | A      |

- **Y / P = 1** (Positive/Yes)
- **N = 0** (Negative/No)
- **A = Missing → Ignored**

---


In [1]:

# Define binary encodings
jack = {'Fever': 1, 'Cough': 0, 'Test-1': 1, 'Test-2': 0, 'Test-3': 0}
mary = {'Fever': 1, 'Cough': 0, 'Test-1': 1, 'Test-2': None, 'Test-3': 1, 'Test-4': 0}
jim  = {'Fever': 1, 'Cough': 1, 'Test-1': 0, 'Test-2': 0, 'Test-3': 0}

# Function to calculate Jaccard Coefficient
def jaccard(dict1, dict2):
    intersect = 0
    union = 0
    for key in dict1:
        if key in dict2 and dict1[key] is not None and dict2[key] is not None:
            if dict1[key] == 1 and dict2[key] == 1:
                intersect += 1
            if dict1[key] == 1 or dict2[key] == 1:
                union += 1
    return intersect / union if union > 0 else 0

# Calculate coefficients
jm = jaccard(jack, mary)
jj = jaccard(jack, jim)
mj = jaccard(jim, mary)

# Display results
print(f"Jaccard Coefficient (Jack, Mary): {jm:.3f}")
print(f"Jaccard Coefficient (Jack, Jim): {jj:.3f}")
print(f"Jaccard Coefficient (Jim, Mary): {mj:.3f}")


Jaccard Coefficient (Jack, Mary): 0.667
Jaccard Coefficient (Jack, Jim): 0.333
Jaccard Coefficient (Jim, Mary): 0.250



## ✅ Results

| Pair         | Jaccard Coefficient |
|--------------|---------------------|
| Jack & Mary  | **0.667**           |
| Jack & Jim   | **0.333**           |
| Jim & Mary   | **0.250**           |

These results show how similar the patients are based on their symptoms and test results. Higher values indicate more similarity based on shared positive attributes.

---
